diff --git a/scientific.cabal b/scientific.cabal
index fc1d88e..99fefeb 100644
--- a/scientific.cabal
+++ b/scientific.cabal
@@ -1,5 +1,5 @@
 name:                scientific
-version:             0.3.4.11
+version:             0.3.4.12
 synopsis:            Numbers represented using scientific notation
 description:
   @Data.Scientific@ provides the number type 'Scientific'. Scientific numbers are
@@ -66,7 +66,7 @@ library
                        Utils
   other-extensions:    DeriveDataTypeable, BangPatterns
   ghc-options:         -Wall
-  build-depends:       base        >= 4.3   && < 4.10
+  build-depends:       base        >= 4.3   && < 4.11
                      , ghc-prim
                      , integer-logarithms >= 1 && <1.1
                      , deepseq     >= 1.3   && < 1.5
