diff --git a/Graphics/Win32/Control.hsc b/Graphics/Win32/Control.hsc
index 6e680ab..4ac8d01 100644
--- a/Graphics/Win32/Control.hsc
+++ b/Graphics/Win32/Control.hsc
@@ -17,6 +17,8 @@
 
 module Graphics.Win32.Control where
 
+##ifndef ghcjs_HOST_OS
+
 import Data.Bits ((.|.))
 import Graphics.Win32.GDI.Types (HMENU, HWND)
 import Graphics.Win32.Message (WindowMessage)
@@ -346,3 +348,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h InitCommonControls"
 -- % , PBM_SETBARCOLOR
 -- % , PBM_SETBKCOLOR
 -- % , PBM_SETRANGE32
+
+##endif
diff --git a/Graphics/Win32/Dialogue.hsc b/Graphics/Win32/Dialogue.hsc
index c80249f..e682ce6 100644
--- a/Graphics/Win32/Dialogue.hsc
+++ b/Graphics/Win32/Dialogue.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.Dialogue where
 
+##ifndef ghcjs_HOST_OS
+
 import Graphics.Win32.GDI.Types
 import Graphics.Win32.Control
 import Graphics.Win32.Message
@@ -334,3 +336,5 @@ foreign import WINDOWS_CCONV "windows.h SetDlgItemTextW"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/GDI/Bitmap.hsc b/Graphics/Win32/GDI/Bitmap.hsc
index 4cfc533..e2e1caa 100644
--- a/Graphics/Win32/GDI/Bitmap.hsc
+++ b/Graphics/Win32/GDI/Bitmap.hsc
@@ -18,6 +18,9 @@
 -----------------------------------------------------------------------------
 
 module Graphics.Win32.GDI.Bitmap(
+##ifdef ghcjs_HOST_OS
+    ) where
+##else
         RasterOp3,
         RasterOp4,
         sRCCOPY,
@@ -421,3 +424,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h CreateDIBitmap"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/GDI/Brush.hsc b/Graphics/Win32/GDI/Brush.hsc
index c6b6709..52444f4 100644
--- a/Graphics/Win32/GDI/Brush.hsc
+++ b/Graphics/Win32/GDI/Brush.hsc
@@ -19,6 +19,7 @@
 
 module Graphics.Win32.GDI.Brush where
 
+##ifndef ghcjs_HOST_OS
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -77,3 +78,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetStockObject"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/GDI/Clip.hsc b/Graphics/Win32/GDI/Clip.hsc
index 56cc527..c220c36 100644
--- a/Graphics/Win32/GDI/Clip.hsc
+++ b/Graphics/Win32/GDI/Clip.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.GDI.Clip where
 
+##ifndef ghcjs_HOST_OS
+
 import Control.Monad
 import Graphics.Win32.GDI.Types
 import System.Win32.Types
@@ -157,3 +159,5 @@ setClipboardViewer wnd =
   failIfNull "SetClipboardViewer" $ c_SetClipboardViewer wnd
 foreign import WINDOWS_CCONV unsafe "windows.h SetClipboardViewer"
   c_SetClipboardViewer :: HWND -> IO HWND
+
+##endif
\ No newline at end of file
diff --git a/Graphics/Win32/GDI/Font.hsc b/Graphics/Win32/GDI/Font.hsc
index 6942c64..dbbd269 100644
--- a/Graphics/Win32/GDI/Font.hsc
+++ b/Graphics/Win32/GDI/Font.hsc
@@ -35,6 +35,8 @@ module Graphics.Win32.GDI.Font
 -}
         where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -207,3 +209,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetStockObject"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
\ No newline at end of file
diff --git a/Graphics/Win32/GDI/Graphics2D.hs b/Graphics/Win32/GDI/Graphics2D.hs
index aaa96a4..42ae043 100644
--- a/Graphics/Win32/GDI/Graphics2D.hs
+++ b/Graphics/Win32/GDI/Graphics2D.hs
@@ -20,6 +20,8 @@
 module Graphics.Win32.GDI.Graphics2D
         where
 
+#ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -226,3 +228,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetTextExtentPoint32W"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+#endif
diff --git a/Graphics/Win32/GDI/HDC.hs b/Graphics/Win32/GDI/HDC.hs
index bee7572..07d77d5 100644
--- a/Graphics/Win32/GDI/HDC.hs
+++ b/Graphics/Win32/GDI/HDC.hs
@@ -21,6 +21,8 @@ module Graphics.Win32.GDI.HDC
         ( module Graphics.Win32.GDI.HDC
         ) where
 
+#ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -320,3 +322,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h DeleteDC"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+#endif
diff --git a/Graphics/Win32/GDI/Palette.hsc b/Graphics/Win32/GDI/Palette.hsc
index e91822c..5a0c616 100644
--- a/Graphics/Win32/GDI/Palette.hsc
+++ b/Graphics/Win32/GDI/Palette.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.GDI.Palette where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -50,3 +52,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h DeleteObject"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/GDI/Path.hs b/Graphics/Win32/GDI/Path.hs
index 388bc6b..62ae1ae 100644
--- a/Graphics/Win32/GDI/Path.hs
+++ b/Graphics/Win32/GDI/Path.hs
@@ -18,6 +18,9 @@
 -----------------------------------------------------------------------------
 
 module Graphics.Win32.GDI.Path
+#ifdef ghcjs_HOST_OS
+    ( ) where
+#else
         ( beginPath, closeFigure, endPath, fillPath, flattenPath
         , pathToRegion, strokeAndFillPath, strokePath, widenPath
         ) where
@@ -91,3 +94,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h WidenPath"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+#endif
\ No newline at end of file
diff --git a/Graphics/Win32/GDI/Pen.hsc b/Graphics/Win32/GDI/Pen.hsc
index c880170..c0dcca0 100644
--- a/Graphics/Win32/GDI/Pen.hsc
+++ b/Graphics/Win32/GDI/Pen.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.GDI.Pen where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -107,3 +109,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h CreatePen"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/GDI/Region.hs b/Graphics/Win32/GDI/Region.hs
index 03a3f38..4e41ec2 100644
--- a/Graphics/Win32/GDI/Region.hs
+++ b/Graphics/Win32/GDI/Region.hs
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.GDI.Region where
 
+#ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import Graphics.Win32.GDI.Types
 
@@ -150,3 +152,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h RectInRegion"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+#endif
diff --git a/Graphics/Win32/GDI/Types.hsc b/Graphics/Win32/GDI/Types.hsc
index c363530..9ecc674 100644
--- a/Graphics/Win32/GDI/Types.hsc
+++ b/Graphics/Win32/GDI/Types.hsc
@@ -54,6 +54,8 @@ module Graphics.Win32.GDI.Types
 -}
         where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 
 import Control.Monad( zipWithM_ )
@@ -395,3 +397,5 @@ foreign import ccall unsafe "HsGDI.h"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
\ No newline at end of file
diff --git a/Graphics/Win32/Icon.hs b/Graphics/Win32/Icon.hs
index be8b84e..286daa5 100644
--- a/Graphics/Win32/Icon.hs
+++ b/Graphics/Win32/Icon.hs
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.Icon where
 
+#ifndef ghcjs_HOST_OS
+
 import Graphics.Win32.GDI.Types
 import System.Win32.Types
 
@@ -49,3 +51,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h DestroyIcon"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+#endif
diff --git a/Graphics/Win32/Key.hsc b/Graphics/Win32/Key.hsc
index 786bff9..9bd7c5d 100644
--- a/Graphics/Win32/Key.hsc
+++ b/Graphics/Win32/Key.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.Key where
 
+##ifndef ghcjs_HOST_OS
+
 import Control.Monad (liftM)
 import Graphics.Win32.GDI.Types (HWND)
 import System.Win32.Types (DWORD, UINT, WORD, ptrToMaybe)
@@ -124,3 +126,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetKBCodePage"
 
 foreign import WINDOWS_CCONV unsafe "windows.h IsWindowEnabled"
   isWindowEnabled :: HWND -> IO Bool
+
+##endif
diff --git a/Graphics/Win32/Menu.hsc b/Graphics/Win32/Menu.hsc
index 45e7a9e..a727d64 100644
--- a/Graphics/Win32/Menu.hsc
+++ b/Graphics/Win32/Menu.hsc
@@ -65,6 +65,8 @@ module Graphics.Win32.Menu
 
        ) -} where
 
+##ifndef ghcjs_HOST_OS
+
 import Graphics.Win32.GDI.Types
 import System.Win32.Types
 
@@ -476,3 +478,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h RemoveMenu"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/Message.hsc b/Graphics/Win32/Message.hsc
index e164ae9..5ffdf3e 100644
--- a/Graphics/Win32/Message.hsc
+++ b/Graphics/Win32/Message.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.Message where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 
 ##include "windows_cconv.h"
@@ -178,3 +180,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h RegisterWindowMessageW"
 ----------------------------------------------------------------
 -- Phew!
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/Misc.hsc b/Graphics/Win32/Misc.hsc
index d2da261..1e9c092 100644
--- a/Graphics/Win32/Misc.hsc
+++ b/Graphics/Win32/Misc.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.Misc where
 
+##ifndef ghcjs_HOST_OS
+
 import Graphics.Win32.GDI.Types
 import System.Win32.Types
 
@@ -298,3 +300,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h timeGetTime"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/Graphics/Win32/Resource.hsc b/Graphics/Win32/Resource.hsc
index e8ad565..57e77a5 100644
--- a/Graphics/Win32/Resource.hsc
+++ b/Graphics/Win32/Resource.hsc
@@ -19,6 +19,8 @@
 
 module Graphics.Win32.Resource where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 
 import Foreign
@@ -148,3 +150,5 @@ updateResource h ty name lang p_data data_len =
     c_UpdateResource h ty c_name lang p_data data_len
 foreign import WINDOWS_CCONV unsafe "windows.h UpdateResourceW"
   c_UpdateResource :: HANDLE -> LPCTSTR -> LPCTSTR -> WORD -> Addr -> DWORD -> IO Bool
+
+##endif
diff --git a/Graphics/Win32/Window.hsc b/Graphics/Win32/Window.hsc
index e5bd152..e0aa17a 100644
--- a/Graphics/Win32/Window.hsc
+++ b/Graphics/Win32/Window.hsc
@@ -18,6 +18,8 @@
 
 module Graphics.Win32.Window where
 
+##ifndef ghcjs_HOST_OS
+
 import Control.Monad (liftM)
 import Data.Maybe (fromMaybe)
 import Data.Word (Word32)
@@ -723,3 +725,5 @@ foreign import WINDOWS_CCONV "windows.h SendMessageW"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/System/Win32/Console.hsc b/System/Win32/Console.hsc
index a40887f..eea5022 100644
--- a/System/Win32/Console.hsc
+++ b/System/Win32/Console.hsc
@@ -18,6 +18,9 @@
 -----------------------------------------------------------------------------
 
 module System.Win32.Console (
+##ifdef ghcjs_HOST_OS
+    ) where
+##else
         -- * Console code pages
         getConsoleCP,
         setConsoleCP,
@@ -60,3 +63,4 @@ foreign import WINDOWS_CCONV safe "windows.h GenerateConsoleCtrlEvent"
     c_GenerateConsoleCtrlEvent :: CtrlEvent -> DWORD -> IO BOOL
 
 -- ToDo: lots more
+##endif
diff --git a/System/Win32/DLL.hsc b/System/Win32/DLL.hsc
index 5416720..50b0bcd 100644
--- a/System/Win32/DLL.hsc
+++ b/System/Win32/DLL.hsc
@@ -19,6 +19,8 @@
 
 module System.Win32.DLL where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 
 import Foreign
@@ -86,3 +88,5 @@ loadLibraryEx name h flags =
   failIfNull "LoadLibraryEx" $ c_LoadLibraryEx c_name h flags
 foreign import WINDOWS_CCONV unsafe "windows.h LoadLibraryExW"
   c_LoadLibraryEx :: LPCTSTR -> HANDLE -> LoadLibraryFlags -> IO HINSTANCE
+
+##endif
diff --git a/System/Win32/DebugApi.hsc b/System/Win32/DebugApi.hsc
index a6ec73f..8ac1627 100644
--- a/System/Win32/DebugApi.hsc
+++ b/System/Win32/DebugApi.hsc
@@ -18,6 +18,8 @@
 -----------------------------------------------------------------------------
 module System.Win32.DebugApi where
 
+##ifndef ghcjs_HOST_OS
+
 import Control.Exception( bracket_ )
 import Data.Word        ( Word8, Word32 )
 import Foreign          ( Ptr, nullPtr, ForeignPtr, mallocForeignPtrBytes
@@ -409,3 +411,5 @@ foreign import WINDOWS_CCONV "windows.h IsDebuggerPresent"
 
 foreign import WINDOWS_CCONV "windows.h  DebugBreak"
     debugBreak :: IO ()
+
+##endif
\ No newline at end of file
diff --git a/System/Win32/File.hsc b/System/Win32/File.hsc
index 920e0d9..10e0408 100644
--- a/System/Win32/File.hsc
+++ b/System/Win32/File.hsc
@@ -26,6 +26,8 @@ module System.Win32.File
 -}
 where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 import System.Win32.Time
 
@@ -655,3 +657,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h SetVolumeLabelW"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/System/Win32/FileMapping.hsc b/System/Win32/FileMapping.hsc
index 9bfd6e4..19c0188 100644
--- a/System/Win32/FileMapping.hsc
+++ b/System/Win32/FileMapping.hsc
@@ -16,6 +16,8 @@
 -----------------------------------------------------------------------------
 module System.Win32.FileMapping where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types   ( HANDLE, DWORD, BOOL, SIZE_T, LPCTSTR, withTString
                             , failIf, failIfNull, DDWORD, ddwordToDwords
                             , iNVALID_HANDLE_VALUE )
@@ -172,3 +174,5 @@ foreign import WINDOWS_CCONV "windows.h UnmapViewOfFile"
 {-# CFILES cbits/HsWin32.c #-}
 foreign import ccall "HsWin32.h &UnmapViewOfFileFinaliser"
     c_UnmapViewOfFileFinaliser :: FunPtr (Ptr a -> IO ())
+
+##endif
diff --git a/System/Win32/Info.hsc b/System/Win32/Info.hsc
index 81da6b2..93381fb 100644
--- a/System/Win32/Info.hsc
+++ b/System/Win32/Info.hsc
@@ -19,6 +19,8 @@
 
 module System.Win32.Info where
 
+##ifndef ghcjs_HOST_OS
+
 import Control.Exception (catch)
 import Foreign.Marshal.Alloc (alloca)
 import Foreign.Marshal.Array (allocaArray)
@@ -394,3 +396,5 @@ type SMSetting = UINT
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+##endif
diff --git a/System/Win32/Mem.hsc b/System/Win32/Mem.hsc
index f3be522..022e1c8 100644
--- a/System/Win32/Mem.hsc
+++ b/System/Win32/Mem.hsc
@@ -19,6 +19,8 @@
 
 module System.Win32.Mem where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types
 
 import Foreign
@@ -317,3 +319,5 @@ virtualUnlock addr size =
   failIfFalse_ "VirtualUnlock" $ c_VirtualUnlock addr size
 foreign import WINDOWS_CCONV unsafe "windows.h VirtualUnlock"
   c_VirtualUnlock :: Addr -> DWORD -> IO Bool
+
+##endif
\ No newline at end of file
diff --git a/System/Win32/NLS.hsc b/System/Win32/NLS.hsc
index eeb3efa..8ef298e 100644
--- a/System/Win32/NLS.hsc
+++ b/System/Win32/NLS.hsc
@@ -16,6 +16,9 @@
 -----------------------------------------------------------------------------
 
 module System.Win32.NLS  (
+##ifdef ghcjs_HOST_OS
+    ) where
+##else
         module System.Win32.NLS,
 
         -- defined in System.Win32.Types
@@ -381,3 +384,4 @@ foreign import WINDOWS_CCONV unsafe "MultiByteToWideChar"
         -> LPWSTR  -- lpWideCharStr
         -> CInt    -- cchWideChar
         -> IO CInt
+##endif
diff --git a/System/Win32/Process.hsc b/System/Win32/Process.hsc
index 7820f0f..2cebeaf 100644
--- a/System/Win32/Process.hsc
+++ b/System/Win32/Process.hsc
@@ -18,6 +18,9 @@
 -----------------------------------------------------------------------------
 
 module System.Win32.Process where
+
+##ifndef ghcjs_HOST_OS
+
 import Control.Exception    ( bracket )
 import Control.Monad        ( liftM5 )
 import Foreign              ( Ptr, peekByteOff, allocaBytes, pokeByteOff
@@ -130,3 +133,5 @@ th32SnapEnumProcesses h = allocaBytes (#size PROCESSENTRY32W) $ \pe -> do
                 entry <- peekProcessEntry32 pe
                 ok' <- c_Process32Next h pe
                 readAndNext ok' pe (entry:res)
+
+##endif
diff --git a/System/Win32/Registry.hsc b/System/Win32/Registry.hsc
index 3c990bb..c0bf182 100644
--- a/System/Win32/Registry.hsc
+++ b/System/Win32/Registry.hsc
@@ -16,6 +16,11 @@
 -----------------------------------------------------------------------------
 
 module System.Win32.Registry
+
+##ifdef ghcjs_HOST_OS
+        () where
+##else
+
                 ( module System.Win32.Registry
                 ) where
 {- What's really on offer:
@@ -523,3 +528,5 @@ regUnLoadKey key subkey =
   failUnlessSuccess "RegUnLoadKey" $ c_RegUnLoadKey p_key c_subkey
 foreign import WINDOWS_CCONV unsafe "windows.h RegUnLoadKeyW"
   c_RegUnLoadKey :: PKEY -> LPCTSTR -> IO ErrCode
+
+##endif
diff --git a/System/Win32/Security.hsc b/System/Win32/Security.hsc
index f00d251..236d0c2 100644
--- a/System/Win32/Security.hsc
+++ b/System/Win32/Security.hsc
@@ -22,7 +22,10 @@
 --
 -----------------------------------------------------------------------------
 
-module System.Win32.Security ( 
+module System.Win32.Security (
+##ifdef ghcjs_HOST_OS
+        ) where
+##else
         -- * Types
         SID, PSID,
         ACL, PACL,
@@ -96,34 +99,34 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorControl"
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> Ptr SECURITY_DESCRIPTOR_CONTROL -- pControl
     -> LPDWORD -- lpdwRevision
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorDacl"
-  c_getSecurityDescriptorDacl 
+  c_getSecurityDescriptorDacl
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> LPBOOL -- lpbDaclPresent
     -> Ptr PACL -- pDacl
     -> LPBOOL -- lpbDaclDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorGroup"
   c_getSecurityDescriptorGroup
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> Ptr PSID -- pGroup
     -> LPBOOL -- lpbGroupDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorLength"
   c_getSecurityDescriptorLength
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
-    -> IO DWORD 
+    -> IO DWORD
 
 foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorOwner"
   c_getSecurityDescriptorOwner
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> Ptr PSID -- pOwner
     -> LPBOOL -- lpbOwnerDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorSacl"
   c_getSecurityDescriptorSacl
@@ -131,18 +134,18 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetSecurityDescriptorSacl"
     -> LPBOOL -- lpbSaclPresent
     -> Ptr PACL -- pSacl
     -> LPBOOL -- lpbSaclDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h InitializeSecurityDescriptor"
   c_initializeSecurityDescriptor
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> DWORD -- dwRevision
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h IsValidSecurityDescriptor"
   c_isValidSecurityDescriptor
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h SetSecurityDescriptorDacl"
   c_setSecurityDescriptorDacl
@@ -150,21 +153,21 @@ foreign import WINDOWS_CCONV unsafe "windows.h SetSecurityDescriptorDacl"
     -> BOOL -- bDaclPresent
     -> PACL -- pDacl
     -> BOOL -- bDaclDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h SetSecurityDescriptorGroup"
   c_setSecurityDescriptorGroup
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> PSID -- pGroup
     -> BOOL -- bGroupDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h SetSecurityDescriptorOwner"
   c_setSecurityDescriptorOwner
     :: PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> PSID -- pOwner
     -> BOOL -- bOwnerDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 foreign import WINDOWS_CCONV unsafe "windows.h SetSecurityDescriptorSacl"
   c_setSecurityDescriptorSacl
@@ -172,7 +175,7 @@ foreign import WINDOWS_CCONV unsafe "windows.h SetSecurityDescriptorSacl"
     -> BOOL -- bSaclPresent
     -> PACL -- pSacl
     -> BOOL -- bSaclDefaulted
-    -> IO BOOL 
+    -> IO BOOL
 
 -- ---------------------------------------------------------------------------
 
@@ -190,7 +193,7 @@ type SECURITY_INFORMATION = DWORD
 -- , pROTECTED_DACL_SECURITY_INFORMATION = PROTECTED_DACL_SECURITY_INFORMATION
 -- , pROTECTED_SACL_SECURITY_INFORMATION = PROTECTED_SACL_SECURITY_INFORMATION
 -- , uNPROTECTED_DACL_SECURITY_INFORMATION = UNPROTECTED_DACL_SECURITY_INFORMATION
--- , uNPROTECTED_SACL_SECURITY_INFORMATION = UNPROTECTED_SACL_SECURITY_INFORMATION 
+-- , uNPROTECTED_SACL_SECURITY_INFORMATION = UNPROTECTED_SACL_SECURITY_INFORMATION
 
 getFileSecurity
     :: String
@@ -203,7 +206,7 @@ getFileSecurity filename si =
   needed <- peek lpnLengthNeeded
   fpSd <- mallocForeignPtrBytes (fromIntegral needed)
   withForeignPtr fpSd $ \pSd -> do
-  failIfFalse_ "getFileSecurity" $ 
+  failIfFalse_ "getFileSecurity" $
     c_GetFileSecurity lpFileName si pSd needed lpnLengthNeeded
   return (SecurityDescriptor fpSd)
 
@@ -214,7 +217,7 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetFileSecurityW"
     -> PSECURITY_DESCRIPTOR -- pSecurityDescriptor
     -> DWORD -- nLength
     -> LPDWORD -- lpnLengthNeeded
-    -> IO BOOL 
+    -> IO BOOL
 
 --foreign import WINDOWS_CCONV unsafe "windows.h AccessCheck"
 --  c_AccessCheck
@@ -235,3 +238,5 @@ foreign import WINDOWS_CCONV unsafe "windows.h GetFileSecurityW"
 --      -> BOOL -- OpenAsSelf
 --      -> PHANDLE -- TokenHandle
 --      -> IO BOOL
+
+##endif
diff --git a/System/Win32/Shell.hsc b/System/Win32/Shell.hsc
index 9dde0a5..90d4390 100644
--- a/System/Win32/Shell.hsc
+++ b/System/Win32/Shell.hsc
@@ -18,6 +18,9 @@
 -----------------------------------------------------------------------------
 
 module System.Win32.Shell (
+##ifdef ghcjs_HOST_OS
+    ) where
+##else
   sHGetFolderPath,
   CSIDL,
   cSIDL_PROFILE,
@@ -81,9 +84,10 @@ sHGetFolderPath hwnd csidl hdl flags =
     peekTString pstr
 
 raiseUnsupported :: String -> IO ()
-raiseUnsupported loc = 
+raiseUnsupported loc =
    ioError (ioeSetErrorString (mkIOError illegalOperationErrorType loc Nothing Nothing) "unsupported operation")
 
 foreign import WINDOWS_CCONV unsafe "SHGetFolderPathW"
   c_SHGetFolderPath :: HWND -> CInt -> HANDLE -> DWORD -> LPTSTR
                     -> IO HRESULT
+##endif
\ No newline at end of file
diff --git a/System/Win32/SimpleMAPI.hsc b/System/Win32/SimpleMAPI.hsc
index c3efec7..119d1e2 100644
--- a/System/Win32/SimpleMAPI.hsc
+++ b/System/Win32/SimpleMAPI.hsc
@@ -19,6 +19,8 @@
 module System.Win32.SimpleMAPI
 where
 
+##ifndef ghcjs_HOST_OS
+
 -- I am not sure why exactly, but with mingw64 mapi.h does not define
 -- some of the values we use, e.g. MAPI_LOGOFF_SHARED.
 -- mapix.h does define MAPI_LOGOFF_SHARED, but the various flags
@@ -217,7 +219,7 @@ mapiLogon f hwnd ses pw flags =
     maybeWith withCAString pw   $ \pw   ->
     alloca                      $ \out  -> do
         mapiFail_ "MAPILogon: " $ mapifLogon
-            f (maybeHWND hwnd) 
+            f (maybeHWND hwnd)
             ses pw flags 0 out
         peek out
 
@@ -415,3 +417,5 @@ mapiSendMail f ses hwnd msg flag = withMessage f ses msg $ \msg ->
 
 handleIOException :: (IOException -> IO a) -> IO a -> IO a
 handleIOException = handle
+
+##endif
diff --git a/System/Win32/Time.hsc b/System/Win32/Time.hsc
index 978a915..5bebcaf 100644
--- a/System/Win32/Time.hsc
+++ b/System/Win32/Time.hsc
@@ -18,6 +18,8 @@
 -----------------------------------------------------------------------------
 module System.Win32.Time where
 
+##ifndef ghcjs_HOST_OS
+
 import System.Win32.Types   ( DWORD, WORD, LONG, BOOL, failIf, failIf_, HANDLE
                             , peekTStringLen, LCID, LPTSTR, LPCTSTR, DDWORD
                             , LARGE_INTEGER, ddwordToDwords, dwordsToDdword )
@@ -309,3 +311,5 @@ getTimeFormat locale flags st fmt =
             size <- failIf (==0) "getTimeFormat: GetTimeFormat" $
                 c_GetTimeFormat locale flags st fmt (castPtr out) size
             peekTStringLen (out,fromIntegral size)
+
+##endif
diff --git a/System/Win32/Types.hs b/System/Win32/Types.hs
old mode 100755
new mode 100644
index 094a594..59d3eb1
--- a/System/Win32/Types.hs
+++ b/System/Win32/Types.hs
@@ -16,6 +16,9 @@
 -----------------------------------------------------------------------------
 
 module System.Win32.Types
+#ifdef ghcjs_HOST_OS
+    ( ) where
+#else
         ( module System.Win32.Types
         , nullPtr
         ) where
@@ -326,3 +329,5 @@ foreign import ccall unsafe "HsWin32.h prim_SUBLANGID"
 ----------------------------------------------------------------
 -- End
 ----------------------------------------------------------------
+
+#endif
